### PR TITLE
perf(flow): Map&Reduce Operator use batch to reduce alloc

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

- add `render_*_batch` functions for flow's render(See `render/map/reduce/src_sink.rs` and `render.rs` files)
 - A major change is impl of `render_reduce_batch` which simpilfy impl and should improve performance(See `src/flow/src/compute/render/reduce.rs`)
- add `evaluate_inner/into_batch` for SafeMfpPlan to help with eval mfp with batch(See `src/flow/src/expr/linear.rs` file)
- other changes are minor and can be ignore

The PR is already too large too add functionality to actually use it, will do that in next PR

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced batch processing capabilities for dataflow plans, enhancing performance and efficiency.
	- Added methods for handling batch operations in various contexts, including rendering plans, processing source data, and managing outputs.
	- Enhanced the `Batch` struct to track row modifications and implemented new methods for improved error handling.

- **Bug Fixes**
	- Improved error handling within batch processing methods to ensure robust operation.

- **Documentation**
	- Updated method documentation for clarity regarding functionality and usage.

- **Tests**
	- New test cases introduced to validate batch processing functionality across various components. 

- **Chores**
	- Added comments for future code quality improvements regarding out-of-bounds access in the codebase.

### CODE REVIEW

**Comment by discord9** [comment]
@coderabbitai review

File: src/flow/src/expr.rs [hunk 1]
```diff
@@ -38,41 +39,165 @@ use snafu::{ensure, ResultExt};
...
+            }
+        );
+        Ok(Self {
+            batch,
+            row_count,
+            diffs: None,
+        })
+    }
+
     pub fn new(batch: Vec<VectorRef>, row_count: usize) -> Self {
```
Comment (zhongzc) [inline src/flow/src/expr.rs:141]:
```suggestion

    pub fn new_unchecked(batch: Vec<VectorRef>, row_count: usize) -> Self {
```

File: src/flow/src/compute/render.rs [hunk 2]
```diff
@@ -112,6 +141,52 @@ impl<'referred, 'df> Context<'referred, 'df> {
...
+    /// render Constant, take all rows that have a timestamp not greater than the current time
+    /// This function is primarily used for testing
+    /// Always assume input is sorted by timestamp
+    pub fn render_constant_batch(&mut self, rows: Vec<DiffRow>) -> CollectionBundle<Batch> {
+        let (send_port, recv_port) = self.df.make_edge::<_, Toff<Batch>>("constant_batch");
+        let mut per_time: BTreeMap<repr::Timestamp, Vec<DiffRow>> = rows
+            .into_iter()
+            .group_by(|(_row, ts, _diff)| *ts)
+            .into_iter()
+            .map(|(k, v)| (k, v.into_iter().collect_vec()))
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:153]: It feels like this line can be removed to reduce memory allocation.

File: src/flow/src/compute/render/map.rs [hunk 3]
```diff
@@ -23,12 +23,58 @@ use crate::compute::render::Context;
...
+        let scheduler = self.compute_state.get_scheduler();
+
+        let subgraph = self.df.add_subgraph_in_out(
+            "mfp_batch",
+            input.collection.into_inner(),
+            out_send_port,
+            move |_ctx, recv, send| {
+                // mfp only need to passively receive updates from recvs
+                let src_data = recv.take_inner().into_iter().flat_map(|v| v.into_iter());
+                let mut output_batches = vec![];
```
Comment (zhongzc) [inline src/flow/src/compute/render/map.rs:57]: with_capacity

File: src/flow/src/compute/render/map.rs [hunk 4]
```diff
@@ -23,12 +23,58 @@ use crate::compute::render::Context;
...
 impl<'referred, 'df> Context<'referred, 'df> {
+    /// Like `render_mfp` but in batch mode
+    pub fn render_mfp_batch(
+        &mut self,
+        input: Box<TypedPlan>,
+        mfp: MapFilterProject,
+    ) -> Result<CollectionBundle<Batch>, Error> {
+        let input = self.render_plan_batch(*input)?;
+
+        let (out_send_port, out_recv_port) = self.df.make_edge::<_, Toff<Batch>>("mfp");
```
Comment (zhongzc) [inline src/flow/src/compute/render/map.rs:40]: It is better to keep L40 and L51 consistent.

File: src/flow/src/compute/render/reduce.rs [hunk 5]
```diff
@@ -14,23 +14,247 @@
...
+
+        let now = self.compute_state.current_time_ref();
+
+        let err_collector = self.err_collector.clone();
+
+        // TODO(discord9): better way to schedule future run
+        let scheduler = self.compute_state.get_scheduler();
+
+        let (out_send_port, out_recv_port) =
+            self.df.make_edge::<_, Toff<Batch>>(Self::REDUCE_BATCH);
```
Comment (zhongzc) [inline src/flow/src/compute/render/reduce.rs:97]: L97 and L101 ditto

**Review by zhongzc — APPROVED** [review]

**Review by waynexia — APPROVED** [review]
General LGTM

**Pull request merged**